// quicktype-core/dist/language/Kotlin/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KotlinXRenderer = exports.KotlinKlaxonRenderer = exports.KotlinJacksonRenderer = exports.KotlinRenderer = exports.kotlinOptions = exports.KotlinTargetLanguage = void 0;
var language_1 = require("./language");
Object.defineProperty(exports, "KotlinTargetLanguage", { enumerable: true, get: function () { return language_1.KotlinTargetLanguage; } });
Object.defineProperty(exports, "kotlinOptions", { enumerable: true, get: function () { return language_1.kotlinOptions; } });
var KotlinRenderer_1 = require("./KotlinRenderer");
Object.defineProperty(exports, "KotlinRenderer", { enumerable: true, get: function () { return KotlinRenderer_1.KotlinRenderer; } });
var KotlinJacksonRenderer_1 = require("./KotlinJacksonRenderer");
Object.defineProperty(exports, "KotlinJacksonRenderer", { enumerable: true, get: function () { return KotlinJacksonRenderer_1.KotlinJacksonRenderer; } });
var KotlinKlaxonRenderer_1 = require("./KotlinKlaxonRenderer");
Object.defineProperty(exports, "KotlinKlaxonRenderer", { enumerable: true, get: function () { return KotlinKlaxonRenderer_1.KotlinKlaxonRenderer; } });
var KotlinXRenderer_1 = require("./KotlinXRenderer");
Object.defineProperty(exports, "KotlinXRenderer", { enumerable: true, get: function () { return KotlinXRenderer_1.KotlinXRenderer; } });
